### PR TITLE
[jit][script] Fix a bug combining sizes/unsized tensors

### PR DESCRIPTION
**Pull request by zdevito**
This add an isSubtypeOf method to reflect that sized tensors are a subtype
of Dynamic[Tensors]. It updates the typechecking code to reflect this
relationship.

### CODE REVIEW

**Comment by zdevito** [comment]
Also adding index_select to shape propagation, because it is needed for onnx export.

**Comment by zdevito** [comment]
@apaszke @jamesr66a

**Comment by jamesr66a** [comment]
you have some stray submodule updates

File: torch/csrc/jit/type.h [hunk 1]
```diff
@@ -207,22 +216,40 @@ struct TupleType : public Type {
     return elements_;
   }
   virtual bool operator==(const Type& rhs) const override {
+    return compare(rhs, [&](const Type& a, const Type& b) {
```
Comment (jamesr66a) [inline torch/csrc/jit/type.h]: Do you need reference capture here?

**Review by jamesr66a — APPROVED** [review]

**Review by apaszke — DISMISSED** [review]
(Not a review yet. Requesting changes just so that no one merges the accidental submodule changes)

**Review by apaszke — COMMENTED** [review]
LGTM, but need a revert for submodules.

**Comment by apaszke** [comment]
Also, where the bug with combining sizes/unsized tensors was?

**Review by apaszke — APPROVED** [review]

**Comment by zdevito** [comment]
The bug was an interaction between TensorType and DynamicType. We'd see TensorTypes in the IR when that graph fragment came from a trace. If you then tried to re-assign one of those values, it would fail because the original type was TensorType and the new type was Dynamic. This PR makes the subtype relationship explicit, and then uses `interpreterType` to wipe away size-specific information  in the type-checker so it would allow this situations.  An alternative, wiping away trace-recorded size information, seemed less robust.

**Pull request merged**